src/libopensc/card-piv.c

```diff
@@ -3312,11 +5184,88 @@ static int piv_init(sc_card_t *card)
 			apdu.resp = yubico_version_buf;
 			apdu.resplen = sizeof(yubico_version_buf);
 			apdu.le = apdu.resplen;
-			r = sc_transmit_apdu(card, &apdu);
-			priv->yubico_version = (yubico_version_buf[0]<<16) | (yubico_version_buf[1] <<8) | yubico_version_buf[2];
-			sc_log(card->ctx, "Yubico card->type=%d, r=0x%08x version=0x%08x", card->type, r, priv->yubico_version);
+			r2 = sc_transmit_apdu(card, &apdu); /* on error yubico_version == 0 */
+			if (r2 >= 3) {
+				priv->yubico_version = (yubico_version_buf[0]<<16) | (yubico_version_buf[1] <<8) | yubico_version_buf[2];
+				sc_log(card->ctx, "Yubico card->type=%d, r=0x%08x version=0x%08x", card->type, r, priv->yubico_version);
+			}
+	}
+	sc_debug(card->ctx,SC_LOG_DEBUG_MATCH, "PIV_MATCH card->type:%d r2:%d CI:%08x r:%d\n", card->type, r2, priv->card_issues, r);
+
+	 /* We now know PIV AID is active, test CCC object. 800-73-* say CCC is required */
+	 /* CCC not readable over contactless, unless using VCI. but don't need CCC for SC_CARD_TYPE_PIV_II_800_73_4 */
+	switch (card->type) {
+		/*
+		 * For cards that may also be CAC, try and read the CCC
+		 * CCC is required and all Dual PIV/CAC will have a CCC
+		 * Currently Dual PIV/CAC are based on NIST 800-73-1 which does not have Discovery or History
+		 */
+		case SC_CARD_TYPE_PIV_II_BASE: /* i.e. really don't know what this is */
+		case SC_CARD_TYPE_PIV_II_GENERIC:
+		case SC_CARD_TYPE_PIV_II_HIST:
+		case SC_CARD_TYPE_PIV_II_GI_DE:
+		case SC_CARD_TYPE_PIV_II_GEMALTO:
+		case SC_CARD_TYPE_PIV_II_OBERTHUR:
+			r2 = piv_process_ccc(card);
+			sc_debug(card->ctx,SC_LOG_DEBUG_MATCH, "PIV_MATCH card->type:%d r2:%d ccc_flags:%08x CI:%08x r:%d\n",
+					card->type, r2, priv->ccc_flags, priv->card_issues, r);
+			/* Ignore any error. */
+			/* If CCC says it has CAC with PKI on card set to one of the SC_CARD_TYPE_PIV_II_*_DUAL_CAC */
+			if (priv->ccc_flags & PIV_CCC_F3_CAC_PKI) {
+				switch (card->type)  {
+					case SC_CARD_TYPE_PIV_II_BASE:
+					case SC_CARD_TYPE_PIV_II_GENERIC:
+					case SC_CARD_TYPE_PIV_II_HIST:
+					case SC_CARD_TYPE_PIV_II_GI_DE:
+					    card->type = SC_CARD_TYPE_PIV_II_GI_DE_DUAL_CAC;
+					    priv->card_issues |= CI_DISCOVERY_USELESS;
+					    priv->obj_cache[PIV_OBJ_DISCOVERY].flags |= PIV_OBJ_CACHE_NOT_PRESENT;
+					    break;
+					case SC_CARD_TYPE_PIV_II_GEMALTO_DUAL_CAC:
```

Comment:
Could this `SC_CARD_TYPE_PIV_II_GEMALTO_DUAL_CAC` and `SC_CARD_TYPE_PIV_II_OBERTHUR_DUAL_CAC` (below) be every matched as we are already in `switch (card->type)` and the only function called in between is `piv_process_ccc`, which does not change `card->type` as far as I see.

Reply:
`SC_CARD_TYPE_PIV_II_GEMALTO_DUAL_CAC` and `SC_CARD_TYPE_PIV_II_OBERTHUR_DUAL_CAC` are not needed in case after the calling `piv_process_ccc` 

A user could have forced the card type to `SC_CARD_TYPE_PIV_II_GEMALTO_DUAL_CAC` or `SC_CARD_TYPE_PIV_II_OBERTHUR_DUAL_CAC`  and that is handled in the next case statement for all 3 CAC type cards. 

The comment" `/*  TODO Move up if user forced it to be one of the CAC types, assume it is CAC */` can also be removed.

Adding fix for this.